keps/sig-node/753-keystone-containers/README.md

```diff
@@ -0,0 +1,946 @@
+
+We are calling `application/essential/main/critical` containers as `keystone` containers in this KEP to distinguish them from sidecars.
+<!--
+This section is incredibly important for producing high-quality, user-focused
+documentation such as release notes or a development roadmap. It should be
+possible to collect this information before implementation begins, in order to
+avoid requiring implementors to split their attention between writing release
+notes and implementing the feature itself. KEP editors and SIG Docs
+should help to ensure that the tone and content of the `Summary` section is
+useful for a wide audience.
+
+A good summary is probably at least a paragraph in length.
+
+Both in this section and below, follow the guidelines of the [documentation
+style guide]. In particular, wrap lines to a reasonable length, to make it
+easier for reviewers to cite specific portions, and to minimize diff churn on
+updates.
+
+[documentation style guide]: https://github.com/kubernetes/community/blob/master/contributors/guide/style-guide.md
+-->
+
+## Motivation
+
+There are lots of [user requests](https://https://github.com/kubernetes/kubernetes/issues/25908) for better support of sidecar containers in Jobs but this issue mostly gets lost in a larger attempt to address the sidecar issue as a whole.
+
+Consider a K8s job with two containers one that does the actual work and another one is for monitoring/logging, the keystone container will complete the work and exit, the sidecar container will keep running and the job will never run to completion. This is problematic for any external application/tool depending on the job waiting to gets completed. Also, the sidecars unnecessarily occupy the system resources.
+
+The only workaround is that the sidecar container needs some kind of signaling from the app container when it exits or the sidecar monitors the app container status. This is not the ideal workaround as it requires containers to be platform aware.
+
+Another motivation for this KEP is to take small steps towards the sidecar containers that will help a lot of users without including complicated logic and changing critical kubelet code paths.
+
+<!--
+This section is for explicitly listing the motivation, goals, and non-goals of
+this KEP.  Describe why the change is important and the benefits to users. The
+motivation section can optionally provide links to [experience reports] to
+demonstrate the interest in a KEP within the wider Kubernetes community.
+
+[experience reports]: https://github.com/golang/go/wiki/ExperienceReports
+-->
+
+### Goals
+- Pod termination by kubelet if application containers have successfully run to completion and exited.
+
+<!--
+List the specific goals of the KEP. What is it trying to achieve? How will we
+know that this has succeeded?
+-->
+
+### Non-Goals
+- Handling Container startup/shutdown ordering for application and sidecar containers.
```

Comment:
IME, startup ordering and shutdown ordering are both required in order to handle the common case of a sidecar proxy, like Istio. It seems like this keystone proposal only solves half the problem, and in a simplistic way (shut down all remaining containers without ordering). If this keystone solution is implemented, and then we later add a startup/shutdown ordering feature, won't the keystone solution then be redundant? Why not just add ordered startup/shutdown instead?

Reply:
Indeed @karlkfi that would have been better for Istio (now that they are supporting sidecar-less deployments too) and I completely understand your frustration.
In fact @rata and others have tried to tackle this ordering for several releases, but they have failed due to a lack of support within the SIG.
As a result, and to avoid such feat, we are taking a baby step approach here, to at least have something merged and start getting feedback from users.

Reply:
Unfortunately, it doesn't look like `type: Keystone` can be easily extended to support startup order.

I agree with @karlkfi that you might need to include this goal or at least lay down how it could be implemented in the future. Baby steps are good, but there needs to be a path.

Reply:
> Unfortunately, it doesn't look like `type: Keystone` can be easily extended to support startup order.

Agree, but the goal is to get feedback on the API to see if that way of labeling containers is convenient, as well as solving this issue with jobs.
That being said, maybe you can suggest edits to the KEP to see if we could turn these types into an ordering.
(I also invite you to check @thockin [proposal](https://docs.google.com/document/d/1Q3685Ic2WV7jPo9vpmirZL1zLVJU91zd3_p_aFDPcS0/edit?usp=sharing) from 2 years ago)

Reply:
In re-familairizing myself with this, it really is mostly the inverse of the previous sidecar KEP, isn't it?  It's simpler in that it doesn't ascribe startup or shutdown ordering, which *does* limit the utility of it.  This KEP alone does not really solve the network-proxy sidecar problems (not completely, anyway).  The question is - does it provide enough value to justify itself for now?

Let's assume that the fundamental problems that have stalled the bigger solution will not go away any time soon.  This KEP, if implemented, could be around for years.  If we leave room, maybe it could grow startup/shutfown ordering semantics.  If we add "real" sequencing (whether that is a DAG or phases or ...), it would obsolete this, but I think we could still define it compatibly - you get to pick one or the other model, but not both.  It's tech-debt that we could probably never drop, but it's not short-term debt.  I would not take this on to accelerate by 1 or 2 releases, but I might take it on to save O(years).

Reply:
I _think_ we can solve those issues for the current proposals that solve the root cause issues sidecars have (Tim's proposal for phases and mine for a DAG).

In the phases proposal, we can say this is the same as being in the "network" phase (or some phase TBD).
In the DAG proposal, we might be able to say this "Provides; Network" or something like that.

I think we might find an "elegant" way to integrate it with those two proposals and not just have some legacy code forever. The question that stays even if we find an elegant way to integrate with those two proposals is: is that enough? As we haven't decided if any of those is our north star, is integrating "nicely" with them enough? Or none of them is our north star and we want something different?

It is definitely hard to make progress when some discussion seem to go on a loop. I think some people want to solve a simpler problem to do baby steps on this, some other want to change the pod startup/shutdown completely to solve ~all the issues and we are not talking with all at the same time, so we find some of other constraints after working quite some time on some path and it is not a very good use of anyone's time.

From what I know we, contributors, are fine solving smaller scope problem or a big scope problem. But it seems like reviewers do not agree between themselves which one is better to solve first, or what requirements any sollution should have. Eg. if we are solving the small scope problem it should have some properties like easily integrate into X or Y proposals, so it is "future proof", etc.

IMHO is understandable for this to happen, as this is a potentially big change, lot of people is involved, lot of thoughts are in the air, etc. so it is hard to coordinate.

Is this my impression or something anyone else feels?

Reply:
It does feel like this is going around in circles endlessly, back in 2018 when we were first discussing this, the idea of a primary container, https://github.com/kubernetes/enhancements/tree/master/keps/sig-node/753-sidecar-containers#mark-one-container-as-the-primary-container was considered, which is very similar to this proposal, but it was decided against as people felt it was too narrow in the problems it solved, and therefore we should implement something that solved a wider set of problems, hence the sidecars approach. But then sidecars was then rejected as it was too complicated/not complicated enough. 

It's also always felt like this feature is not that high priority by the SIGs that need to support it, though there are some people in k8s like Thockin who have always tried to move this forward. I think relying on some community members alone without the full backing of sig-node is always going to be extremely difficult for such a big feature.

Reply:
Take my PoV with a grain-of-salt. I'm not a sig-node reviewer and I'm not familiar with the entire history of the proposal. But I'm an API reviewer for sig-apps and sig-scheduling.

My main take-away is that you should do small steps, but plan for extension.

I like the lifecycle struct, but I don't like the `type` field. As already stated, there are two parts of the problem: start up and tear down order. This KEP seems to be focusing on tear down, but in a very limiting way. Maybe you could focus on the entire problem of tear down.

Something like:

```yaml
lifecycle:
  termination:
    onContainersTermination: ['main']
```

Probably worth looking at existing solutions to gather ideas for APIs: https://github.com/karlkfi/kubexit and https://github.com/redboxllc/scuttle

As for gathering backing from SIG-node, probably the best avenue is to consistently discuss the topic in SIG meetings, or asking for a single approver as point-of-contact.

Reply:
@matthyx thinking about it, introducing the field `type` put a lot of pressure on this KEP in terms of future thinking and startup ordering. What if instead we change the KEP to introduce per-container restart policies and include the policy to fail the entire pod?

- pod restart policies are: `Always`, `OnFailure`, `Never`
- new: container restart policy override: `Always`, `OnFailure`, `Never`, `TerminatePod`

This will also allow to restart sidecars for long running jobs without interrupting those.

Reply:
_I_ like it, but it may suffer the same issue @thockin mentions (it may become legacy code once sidecar KEP that adress the rest of the challenges is merged).

I do think that this, though, might not become legacy code, though. This is not something that any of the proposals (phases, the DAG) will allow to model, and it is something useful and that was requested (I think the term used in the past was "critical containers", that has more use cases).

So, for me, this seems like a nice split of the problem space and it won't become legacy code, as it is an split with no overlap with the rest of the proposals. But I'd like to see what others, in particular @thockin thinks, as this is not something easy to consider and weight.

Reply:
@SergeyKanzhelev I like your idea :)
I need to check how it works implementation-wise, but at least I could modify the KEP to add this alternative API.
@thockin thoughts?

Reply:
I wrote it down with more details: https://docs.google.com/document/d/1gX_SOZXCNMcIe9d8CkekiT0GU9htH_my8OB7HSm3fM8/edit#

Reply:
This is great! It's an elegant way to solve the job problem, and it won't interfere with the "real" sidecar KEP.
Should I modify the current KEP, or you (@SergeyKanzhelev ) will fill a new one to supersede that one?

Reply:
the per container override of a restart policy seems promising.

will review the linked google doc, thanks for the suggestion @SergeyKanzhelev

Reply:
> Should I modify the current KEP, or you (@SergeyKanzhelev ) will fill a new one to supersede that one?

@matthyx if you are ok to keep driving it, it would be best if you can start a new KEP from the document. This way we preserve comments for the keystone part of the discussion and separate them from restart policy.

Reply:
I have spent a good part of the last few days thinking on this, and I do NOT think this approach works.  I've tried to distill my thinking into pictures:

https://docs.google.com/presentation/d/1E-PrQtOY8RgcDRJ9HbGDCm9gyMODRRusaatJFEa2vnY/edit#slide=id.gff598714d9_0_21

I don't know if that helps, but I think it is cleaner and more future-safe.

Reply:
access denied

Reply:
@rata 

1) putting container-termination logic (kill helpers when the app completes) behind a field that is named "restart policy" is a very non-obvious API.
2) the described semantics only allow one such container
3) this model requires sidecar injectors to modify stanzas of the pod other than their own
4) this model ONLY describes end-of-life; if we want to extend to startup sequencing it would need another field
5) this API enables MUCH more than we need for the problem at hand and, while I WANT container restart, we have not done the work to fully thing thru all the new corner cases.

I wrote about this a bunch in Sergey's doc (comments, sorry).  As an API, this will not fly.

I drew my pictures as a riff on Derek's suggestion of systemd BindTo.  Again, purely from an API point of view, it's problematic to name specific containers (requires more info than you want) and forces injectors to use heuristics.  I think we get pretty much ALL of the use cases with none of the downsides by indirecting through some per-container metadata.  I tried to walk readers into that conclusion, but obviously did not succeed.

A DAG is a terrible API unless you need a DAG.  I don't think we do.

I'm happy to jump on a Meet or Zoom and discuss.  Time is running VERY short (again) and I have not heard back from Derek in a few days :(

> perhaps a better way to solve these requirements would be to create a new serviceContainers model, that like initContainers have their own separate interface, lifecycle, behavior, and requirements.

This is ... not terrible.  It suffers one big problem, which maybe we don't care about - some specific cases, like network proxies REALLY do need an X before Y guarantee.  So now you end up with containers, initContainers, serviceContainers, and initServiceContainers.  This is more or less what I described as phases or runlevels.

BTW I like "runlevel" because it evokes the init systems BEFORE systemd, which were generally simpler (and less powerful).  I don't think we need an explicit DAG inside a pod.

Reply:
@thockin you have very specific requirements and scenarios in mind. And if we all can agree with those, I think it will be great to proceed with "sidecar", "helpers", "ambient" or whatever other name can be used. But as last time the proposal hits the wall of a decision that SIG Node needs to make: what is the ultimate flexibility we will allow in configuring containers interdependencies. Is it runlevels, is it DAG, is it non-declarative hooks-based API that user can implement?

The proposal of `restartPolicy` and `TerminatePod` is an attempt to suggest the minimal set of fields that can likely fit in most models of configuring containers interdependencies. The idea is to define fields describing specific characteristics of containers rather than their "class", that implies a number of these characteristics. Since those are just characteristics, they can be enforced on API level later for certain types of containers, e.g. sidecars MUST declare `restartPolicy=always`. And yes, the proposal will not solve the automatic injection or startup sequence.

As for specific API - TerminatePod can be a separate property from `restartPolicy` override. It was just very logical in my head to have a policy that says - when you done restarting, make the pod exit. It is still a policy and it is about restart. It also has a benefit of forcing all TerminatePod containers to match the restartPolicy to the Pod's restart policy. Making these containers more predictable for existing controllers.

For the corner cases, the current proposal is limited enough to have a minimal impact on corner cases.

This all said explaining the proposal, having `runlevel=sidecar` move forward will be a great step forward.

I also wrote down more thoughts here: https://docs.google.com/document/d/10xqIf19HPPUa58GXtKGO7D1LGX3t4wfmqiJUN4PJfSA/edit#heading=h.a3qp744ia2p6

Reply:
> The proposal of restartPolicy and TerminatePod

I wanted to be sure I was not way out of line here, so I solicited a consultation from another seasoned kube API reviewer.  Their reaction was the same as mine wrt the use of a "restart policy" to mean "termination policy".  They suggested instead, effectively BindTo.  I asked the same questions I asked here and in the doc - how many can you have, is it OK for injectors to "chain", etc, and they proposed a numerical set of stages.  I described the "runlevels" (they did not like the name :) approach which was met with "yeah, that's better".

So I do not think I am off target.  As far as APIs go, `restartPolicy` is at the bottom of the list.

What we are really trying to do here is put a stake in the ground.  Whatever we decide here is ALPHA.  We can always change it later.  Seriously, if want to alpha it as any of the following, I will OK it for API:

* `sidecar bool` - default false, set true on helper containers
* `keystone bool` - default true, set false on helper containers
* `runlevel PodRunlevel` - default to "App", set to "Sidecar" or "Helper" or ... on sidecars
* any other name that means roughly the same as the above.

Obviously, I think an enum-string is the better choice.

We need someone from sig-node who can approve a tentative, cautious, minimalist step forward.  @derekwaynecarr @mrunalp @dchen1107 @sjenning ?

Reply:
> We can always change it later.

Can we? The field is going into the PodSpec.

Reply:
I would love to get more involved in this KEP as this is an issue we’re running into right now. We’ve been forced to use the Lyft patch for sidecars but I would love to get back on vanilla K8s to make our upgrade process much simpler.

I had a couple questions surrounding the different proposed methods (keystone, container lifecycle, run levels). I’ve read most, if not all, the discussion material as well as the KEP, so I feel up to date on everything.

For container lifecycle:

- If the pod restartPolicy is `Always` and the container override is set to `TerminatePod`, is that a valid scenario? Does `TerminatePod` ever get exercised in that case?

I agree with @thockin that `restartPolicyOverride` is a bit confusing when used with `TerminatePod` since it’s not really a restart policy but more of a question of “what should happen to the pod if this container terminates?”

I do feel like the `restartPolicyOverride` is an interesting feature that could complement keystone containers (minus `TerminatePod` since the `type: Keystone` field essentially does the same as `restartPolicyOverride: TerminatePod`).

For runlevels:

- If I understand correctly, the run levels are ordered, but you can’t necessarily order containers within each runlevel, right?

- If that is the case, do runlevels solve the issue for Istio/network proxies? If I understand correctly, network proxy sidecars need to run before any other containers in order to setup the network. In which case, the helper runlevel would need to run before the init containers as well as all the other containers in the helper runlevel, correct?

 Overall, I do like the run-level approach as it seems the most extendable, though I agree the name is a bit confusing. “Phases” seems more understandable, I remember reading that there might already be a KEP for phases? Is that KEP essentially the same thing as run-levels?

Though I also echo the thoughts by @alculquicondor . Although this feature is alpha, once we include anything in the podspec, wouldn’t it be difficult to change/remove since k8s has never versioned the podspec? Adding something and then changing/removing it wouldn’t be backwards compatible without versioning. And I imagine versioning the podspec would be a big ask.

Reply:
> I do feel like the restartPolicyOverride is an interesting feature that could complement keystone containers (minus TerminatePod since the type: Keystone field essentially does the same as restartPolicyOverride: TerminatePod).

Agree, I think that is a different KEP (or KEPs), though.

> If I understand correctly, the run levels are ordered, but you can’t necessarily order containers within each runlevel, right?

We don't need to guarantee it, though practically speaking they are started in definition order.  This is an implementation detail that has probably escaped far enough that we should codify it.  The alternative would be to spec it as undefined ordering and then randomize startup order within a runlevel.  That seems...hostile?  Sort of like if Go randomized `map` accesses..oh...

> do runlevels solve the issue for Istio/network proxies?

I think runlevels DO NOT solve for istio cases.  I suspect we need more explicit support for that - perhaps another "pre-init" runlevel or something.  I am proposing that we DO NOT solve for that now.  The runlevel proposal makes this use case better in many cases and no worse in any case (I think?).  We can solve for it in beta or even after, once we know if the model has legs or not.

> I agree the name is a bit confusing. “Phases” seems more understandable

I originally called it "phases" but we already have a notion of pod "phase", and it does not mean the same thing.  I am IN NO WAY attached to the words "runlevel", "helper", "sidecar" etc.  :)  I suck at naming things, everyone should know that by now.

> Although this feature is alpha, once we include anything in the podspec, wouldn’t it be difficult to change/remove since k8s has never versioned the podspec?

We have precedent for removing alpha fields from GA APIs.  To use it during alpha, you have to go out of your way to enable a feature gate that says "alpha", so I don't feel (too) bad taking it away.

Reply:
I caught up with @dchen1107 today and with @SergeyKanzhelev and @mrunalp and walked them through a revised proposal deck (https://docs.google.com/presentation/d/1E-PrQtOY8RgcDRJ9HbGDCm9gyMODRRusaatJFEa2vnY/edit).  I think I cajoled them all into seeing my point (I hope nobody feels run over, I know I can be a bit pushy).

There are good questions, like:
* What do we expect to happen with a logging sidecar and a network sidecar?  Does the logger catch the proxy's logs?  Does the proxy catch the logger's net traffic?
* What do we need to know before this can beta?
* Can it be a container field, or does it need to be another array of Container (like initContainers) to handle version skew (old kubelets)?  That makes me sad, but we need to consider it.

We have approximately 24.5 hours until deadline.  Sergey and I will try to sketch something up tomorrow, but it will be tight.  We may not make the deadline, but it won't be for lack of trying.

Reply:
@adisky and @matthyx I don't want to steal this from you, I am just trying like mad to "unstick" it.

Reply:
No problem (at least for me) Tim... being in another timezone I just keep catching up on all these messages during my lunch breaks.
I really appreciate your efforts, as well as those from Dawn, Sergey, Mrunal and all the others participating in these talks.

Reply:
I should also be clear - the concerns that are causing the stuck-ness are legitimate and real and hard to answer.  I don't mean to imply otherwise.

Reply:
What do we need in order to get this merged before the 1.26 freeze?

Reply:
PRR is approved, we just need approve from sig-node leads (@dchen1107 or @derekwaynecarr)... @thockin do you think we should merge the KEP to have it tracked and continue debating after the freeze?

Reply:
So I'm assuming the 1.26 freeze has already started, but we should be able to merge this in afterwards for the 1.27 release? I'm not very familiar with the K8s release process. @thockin do you know when would the 1.27 release likely be? And if this were to get merged soon after the 1.26 freeze, would the 1.27 release need to come out first before any implementation could occur?

Reply:
I'm going to post a top-level comment to reply to these, so it doesn't get lost on any new push :)